File: swift/0110-balanced-binary-tree.swift

class Solution {
    func getMaxHeight(of root: TreeNode?) -> Int {
        guard let root = root else { return 0 }
        return 1 + max(
            getMaxHeight(of: root.left),
            getMaxHeight(of: root.right)
        )
    }
    
    func isBalanced(_ root: TreeNode?) -> Bool {
        guard let root = root else { return true }
        
        let leftHeight  = getMaxHeight(of: root.left)
        let rightHeight = getMaxHeight(of: root.right)
        let difference  = abs(leftHeight - rightHeight)    
    
        // early exits
        guard difference <= 1 else { return false }
        guard isBalanced(root.left)  == true else { return false }
        guard isBalanced(root.right) == true else { return false }
        
        return true
    }
}

File: kotlin/0110-balanced-binary-tree.kt
class Solution {
    fun isBalanced(root: TreeNode?): Boolean {
        return isBalancedUtil(root) != Int.MAX_VALUE
    }

    fun isBalancedUtil(node: TreeNode?): Int {
        if (node == null)
            return 0

        val left = isBalancedUtil(node.left)
        val right = isBalancedUtil(node.right)

        if (left == Int.MAX_VALUE || right == Int.MAX_VALUE)
            return Int.MAX_VALUE

        if (Math.abs(left - right) > 1)
            return Int.MAX_VALUE

        return Math.max(left, right) + 1
    }
}
